Make HDF5 an imported target in CMake

nilsvu opened the pull request:
## Proposed changes

Make HDF5 an imported target in CMake

Instead of linking a library into all of our targets using the
`SPECTRE_LIBRARIES` list we can define "imported targets" for
our dependencies and selectively link them into our targets.
This commit presents an example for HDF5. It is needed only
within the implementation of our `IO` library, so it is linked with the
`PRIVATE` keyword.

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [ ] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

kidder commented:
labeling `in progress` as it doesn't work...

nilsdeppe commented on src/IO/CMakeLists.txt:
```diff
@@ -24,9 +24,12 @@ add_spectre_library(${LIBRARY} ${LIBRARY_SOURCES})
-  INTERFACE DataStructures
-  INTERFACE ErrorHandling
-  INTERFACE Utilities
+  PRIVATE
```
I'm still pretty sure interface is the right call here...

nilsvu replied:
I made all of these libraries `PUBLIC` because I think they are all used both in source and header files. Do you agree this is the right thing to do?

nilsdeppe commented on src/IO/CMakeLists.txt:
```diff
@@ -24,9 +24,12 @@ add_spectre_library(${LIBRARY} ${LIBRARY_SOURCES})
-  INTERFACE Utilities
+  PRIVATE
+    ErrorHandling
+    Hdf5
```
Have you tried just `INTERFACE ${HDF5_LIBRARIES}`? That seems easier

nilsvu commented:
Yes this should be labeled in progress for now :)

Here's some information for using the `PRIVATE`, `PUBLIC` and `INTERFACE` keywords:
- https://cmake.org/pipermail/cmake/2016-May/063400.html
- https://stackoverflow.com/questions/26037954/cmake-target-link-libraries-interface-dependencies

Here's a blog post that I found useful:
- https://pabloariasal.github.io/2018/02/19/its-time-to-do-cmake-right/

Quoting from the last link:
> **Private** properties are used internally to build the target, while **interface** properties are used externally by users of the target.

I think we have neglected the right use of these keywords in code review so far, and we could start paying some attention to them going forward.

wthrowe commented:
@nilsleiffischer The first link you posted is a 404.

nilsvu commented:
@wthrowe sorry, fixed

nilsvu committed:
Make HDF5 an imported target in CMake

Instead of linking a library into all of our targets using the
SPECTRE_LIBRARIES list we can define "imported targets" for
our dependencies and selectively link them into our targets.
This commit presents an example for HDF5, which only needs to
be linked in our IO library.

nilsdeppe commented on src/IO/CMakeLists.txt:
```diff
@@ -24,9 +24,11 @@ add_spectre_library(${LIBRARY} ${LIBRARY_SOURCES})
-  INTERFACE DataStructures
-  INTERFACE ErrorHandling
-  INTERFACE Utilities
+  PUBLIC
```
From what I've read in the links you sent me, it seems like `Hdf5` and `ErrorHandling` should be private. `Utilities` in theory should be private, but I think because `TaggedTuple` is still there it might not be.

nilsvu replied:
Yea I would like to make `Hdf5` and `ErrorHandling` private, but some header files in `IO` include headers from those libraries and therefore they must probably be public, don't you think?

nilsdeppe replied:
Oh yea, you're right... We can see if that can be fixed at some point in the future :)

nilsdeppe commented on cmake/SpectreAddTestLibs.cmake:
```diff
@@ -91,10 +91,9 @@ function(add_test_library LIBRARY FOLDER LIBRARY_SOURCES LINK_LIBS)
-    INTERFACE
-    ${PYTHON_LIBRARIES}
+    PRIVATE
+      ${LINK_LIBS}
```
General comment, please don't indent these more than what `PRIVATE`,etc. are already indented

nilsvu replied:
do you really prefer that? Perhaps we should add this indentation while updating the linking?

nilsdeppe replied:
Yes, no extra indentation is consistent with how we've been doing all of our CMake formatting. No, I don't want the extra spaces. Currently we've been using the formatting Emacs provides, which is not the extra indentation.

nilsvu replied:
fine with me. rebased and squashed.

nilsdeppe replied:
Just a ping on this file :) The rest look good!

nilsdeppe reviewed (commented):
I'm happy for you to squash everything

nilsdeppe reviewed (approved):
LGTM!

![image](https://user-images.githubusercontent.com/13531030/76096878-ba15a280-5f94-11ea-8705-512347bb60a4.png)
